In [1]:
!pip install ucimlrepo

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
!ls "/content/drive/MyDrive/WESAD.zip"

/content/drive/MyDrive/WESAD.zip


In [4]:
print('Listing contents of MyDrive:')
!ls "/content/drive/MyDrive"

Listing contents of MyDrive:
'15).xlsx'
'2014 Izamar Cardenas Resume.gdoc'
'2014 Tax Return Documents (DOMINGO MICHAEL R).pdf'
'2015 Tax Return Documents (DOMINGO MICHAEL R).pdf'
'2016_Domingo_Resume (1).pdf'
 2016_Domingo_Resume.pdf
'2016 State Return (Michael Domingo).pdf'
 20200304163904.pdf
'20250116 2h2'
'20250116 2h2 (1)'
'2025 Camo to Careers - LinkedIn Coaches Event.pdf'
'30 60 90 day Action Plan.docx.gdoc'
 8111-50-26.pdf
'Business letter (1).gdoc'
'Business letter.gdoc'
'Colab Notebooks'
'Comprehensive SAM.gdoc'
'Copy of 20250116 2h2'
'Copy of Group 2 Technical Report.gdoc'
'Copy of Michael Domingo - ApplyAll Applications.gsheet'
'Copy of RO Copy of Japan Itinerary   Guide.gdoc'
'Copy of Veterans Day Minitheme by Slidesgo.gslides'
'CrossFit 2016 open'
'Diabetes Prediction Using CDC Health Indicators (1).gslides'
'Diabetes Prediction Using CDC Health Indicators.gslides'
 domingo.pdf
 emailreceipt_20170928R0945961941.pdf
'Final_Clothing Classification and Clustering Using Deep 

If your `WESAD.zip` file is inside a subfolder, for example, named `MyProject`, you would then run:

```python
!ls "/content/drive/MyDrive/MyProject"
```

Once you find the file, use the full path in your code.

In [12]:
from google.colab import drive
import os

# 1. Mount the drive
drive.mount('/content/drive')

# 2. Define the exact path (Google Drive root is 'MyDrive')
zip_path = '/content/drive/MyDrive/WESAD.zip'

# 3. Check if the file exists before unzipping
if os.path.exists(zip_path):
    print("Found it! Unzipping now...")
    !unzip -q "{zip_path}" -d "/content/WESAD_data"
    print("Done! Files are now in the 'WESAD_data' folder on the left sidebar.")
else:
    print("Still can't see it. Make sure the file is in the main 'My Drive' folder, not a subfolder.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Found it! Unzipping now...
replace /content/WESAD_data/WESAD/S10/S10.pkl? [y]es, [n]o, [A]ll, [N]one, [r]ename: A
Done! Files are now in the 'WESAD_data' folder on the left sidebar.


# Task
Explore the structure of the unzipped WESAD data by listing the contents of the `/content/WESAD_data` directory to identify the main data files.

## Explore Unzipped Data

### Subtask:
Explore the structure of the unzipped WESAD data by listing the contents of the `/content/WESAD_data` directory to identify the main data files.


**Reasoning**:
To explore the structure of the unzipped WESAD data, I will list the contents of the `/content/WESAD_data` directory using the `!ls -F` command as instructed.



In [6]:
print('Listing contents of the unzipped WESAD data directory:')
!ls -F /content/WESAD_data

Listing contents of the unzipped WESAD data directory:
WESAD/


**Reasoning**:
Since the previous `ls` command revealed a subdirectory named `WESAD/`, I will now list the contents of this subdirectory to further explore the data structure and locate the actual data files.



In [7]:
print('Listing contents of the WESAD subdirectory:')
!ls -F /content/WESAD_data/WESAD

Listing contents of the WESAD subdirectory:
S10/  S13/  S15/  S17/	S3/  S5/  S7/  S9/
S11/  S14/  S16/  S2/	S4/  S6/  S8/  wesad_readme.pdf


**Reasoning**:
The previous command revealed subject-specific subdirectories (S2/, S3/, etc.) and a `wesad_readme.pdf` file. To understand the data structure further, I will list the contents of one of these subject directories, specifically `S2/`, to see what kind of data files are stored there.



In [8]:
print('Listing contents of the S2 subdirectory:')
!ls -F /content/WESAD_data/WESAD/S2

Listing contents of the S2 subdirectory:
S2_E4_Data.zip	S2.pkl	S2_quest.csv  S2_readme.txt  S2_respiban.txt


## Load Data

### Subtask:
Load the relevant WESAD data files into pandas DataFrames for cleaning.


**Reasoning**:
I will import the pandas library, define the base path to the S2 subject data, and then load the 'S2.pkl' and 'S2_quest.csv' files into two separate pandas DataFrames, `df_s2_data` and `df_s2_quest` respectively, to begin the data loading process as instructed.



In [13]:
import pandas as pd

# Define the base path to the S2 subject data directory
s2_data_path = '/content/WESAD_data/WESAD/S2'

# Load the S2.pkl file into a pandas DataFrame
df_s2_data = pd.read_pickle(f'{s2_data_path}/S2.pkl')

# Load the S2_quest.csv file into a pandas DataFrame
df_s2_quest = pd.read_csv(f'{s2_data_path}/S2_quest.csv')

print("df_s2_data head:")
print(df_s2_data.head())

print("\ndf_s2_quest head:")
print(df_s2_quest.head())

df_s2_data head:


AttributeError: 'dict' object has no attribute 'head'

**Reasoning**:
The previous code failed because `df_s2_data` was a dictionary, not a pandas DataFrame, and therefore did not have a `.head()` method. The `S2.pkl` file, when loaded, is a dictionary containing various data points. To address the subtask of loading relevant data into DataFrames, I will first load the `.pkl` file into a dictionary, then extract the 'chest' and 'wrist' signal data from it and convert them into separate pandas DataFrames. I will also correctly load the `S2_quest.csv` into a DataFrame.



In [14]:
import pandas as pd

# Define the base path to the S2 subject data directory
s2_data_path = '/content/WESAD_data/WESAD/S2'

# Load the S2.pkl file into a dictionary first
s2_raw_data_dict = pd.read_pickle(f'{s2_data_path}/S2.pkl')

# Extract 'chest' and 'wrist' signals and convert them to DataFrames
df_s2_chest_data = pd.DataFrame(s2_raw_data_dict['signal']['chest'])
df_s2_wrist_data = pd.DataFrame(s2_raw_data_dict['signal']['wrist'])

# Load the S2_quest.csv file into a pandas DataFrame
df_s2_quest = pd.read_csv(f'{s2_data_path}/S2_quest.csv')

print("df_s2_chest_data head:")
print(df_s2_chest_data.head())

print("\ndf_s2_wrist_data head:")
print(df_s2_wrist_data.head())

print("\ndf_s2_quest head:")
print(df_s2_quest.head())

ValueError: Per-column arrays must each be 1-dimensional

**Reasoning**:
The previous code failed because the 'ACC' (accelerometer) data within the 'chest' and 'wrist' signals of the `S2.pkl` file are multi-dimensional arrays, causing a `ValueError` when `pd.DataFrame` tried to interpret them as single columns. To fix this, I need to explicitly extract the 'ACC' data into separate columns (e.g., 'ACC_x', 'ACC_y', 'ACC_z') and then create DataFrames from all the extracted 1-dimensional signal arrays.



In [15]:
import pandas as pd
import numpy as np

# Define the base path to the S2 subject data directory
s2_data_path = '/content/WESAD_data/WESAD/S2'

# Load the S2.pkl file into a dictionary first
s2_raw_data_dict = pd.read_pickle(f'{s2_data_path}/S2.pkl')

# Process 'chest' signals
chest_signals = s2_raw_data_dict['signal']['chest']
chest_data = {}
for sensor_name, data_array in chest_signals.items():
    if sensor_name == 'ACC':
        # Accelerometer data is typically 3-axis, so create three columns
        chest_data['ACC_x'] = data_array[:, 0]
        chest_data['ACC_y'] = data_array[:, 1]
        chest_data['ACC_z'] = data_array[:, 2]
    else:
        # Other sensors are typically 1D
        chest_data[sensor_name] = data_array

df_s2_chest_data = pd.DataFrame(chest_data)

# Process 'wrist' signals
wrist_signals = s2_raw_data_dict['signal']['wrist']
wrist_data = {}
for sensor_name, data_array in wrist_signals.items():
    if sensor_name == 'ACC':
        # Accelerometer data is typically 3-axis, so create three columns
        wrist_data['ACC_x'] = data_array[:, 0]
        wrist_data['ACC_y'] = data_array[:, 1]
        wrist_data['ACC_z'] = data_array[:, 2]
    else:
        # Other sensors are typically 1D
        wrist_data[sensor_name] = data_array

df_s2_wrist_data = pd.DataFrame(wrist_data)

# Load the S2_quest.csv file into a pandas DataFrame
df_s2_quest = pd.read_csv(f'{s2_data_path}/S2_quest.csv')

print("df_s2_chest_data head:")
print(df_s2_chest_data.head())

print("\ndf_s2_wrist_data head:")
print(df_s2_wrist_data.head())

print("\ndf_s2_quest head:")
print(df_s2_quest.head())

ValueError: Per-column arrays must each be 1-dimensional

**Reasoning**:
The previous code failed because some non-accelerometer sensor data (e.g., ECG, EMG) were still 2-dimensional arrays with a single column, which `pd.DataFrame` cannot directly handle as individual columns. I need to flatten these 2D arrays into 1-dimensional arrays before adding them to the `chest_data` and `wrist_data` dictionaries to resolve the `ValueError`.



In [16]:
import pandas as pd
import numpy as np

# Define the base path to the S2 subject data directory
s2_data_path = '/content/WESAD_data/WESAD/S2'

# Load the S2.pkl file into a dictionary first
s2_raw_data_dict = pd.read_pickle(f'{s2_data_path}/S2.pkl')

# Process 'chest' signals
chest_signals = s2_raw_data_dict['signal']['chest']
chest_data = {}
for sensor_name, data_array in chest_signals.items():
    if sensor_name == 'ACC':
        # Accelerometer data is typically 3-axis, so create three columns
        chest_data['ACC_x'] = data_array[:, 0]
        chest_data['ACC_y'] = data_array[:, 1]
        chest_data['ACC_z'] = data_array[:, 2]
    else:
        # Other sensors might be 2D arrays with a single column, flatten them
        chest_data[sensor_name] = data_array.flatten()

df_s2_chest_data = pd.DataFrame(chest_data)

# Process 'wrist' signals
wrist_signals = s2_raw_data_dict['signal']['wrist']
wrist_data = {}
for sensor_name, data_array in wrist_signals.items():
    if sensor_name == 'ACC':
        # Accelerometer data is typically 3-axis, so create three columns
        wrist_data['ACC_x'] = data_array[:, 0]
        wrist_data['ACC_y'] = data_array[:, 1]
        wrist_data['ACC_z'] = data_array[:, 2]
    else:
        # Other sensors might be 2D arrays with a single column, flatten them
        wrist_data[sensor_name] = data_array.flatten()

df_s2_wrist_data = pd.DataFrame(wrist_data)

# Load the S2_quest.csv file into a pandas DataFrame
df_s2_quest = pd.read_csv(f'{s2_data_path}/S2_quest.csv')

print("df_s2_chest_data head:")
print(df_s2_chest_data.head())

print("\ndf_s2_wrist_data head:")
print(df_s2_wrist_data.head())

print("\ndf_s2_quest head:")
print(df_s2_quest.head())

ValueError: All arrays must be of the same length

**Reasoning**:
The previous code failed because sensor data arrays within 'wrist' (and likely 'chest') have different lengths due to varying sampling rates, making it impossible to directly create a single pandas DataFrame. To resolve the `ValueError: All arrays must be of the same length`, I will create separate DataFrames for each individual sensor signal (e.g., 'ACC', 'BVP', 'EDA', 'TEMP', 'ECG', 'EMG') and store them in dictionaries (`chest_dataframes` and `wrist_dataframes`) for organized access.



In [17]:
import pandas as pd
import numpy as np

# Define the base path to the S2 subject data directory
s2_data_path = '/content/WESAD_data/WESAD/S2'

# Load the S2.pkl file into a dictionary first
s2_raw_data_dict = pd.read_pickle(f'{s2_data_path}/S2.pkl')

# Initialize dictionaries to hold individual sensor DataFrames
chest_dataframes = {}
wrist_dataframes = {}

# Process 'chest' signals
chest_signals = s2_raw_data_dict['signal']['chest']
for sensor_name, data_array in chest_signals.items():
    if sensor_name == 'ACC':
        # Accelerometer data is 3-axis, create three columns
        chest_dataframes[sensor_name] = pd.DataFrame(data_array, columns=['ACC_x', 'ACC_y', 'ACC_z'])
    else:
        # Other sensors are typically 1D or 2D with one column, flatten and create a DataFrame
        chest_dataframes[sensor_name] = pd.DataFrame(data_array.flatten(), columns=[sensor_name])

# Process 'wrist' signals
wrist_signals = s2_raw_data_dict['signal']['wrist']
for sensor_name, data_array in wrist_signals.items():
    if sensor_name == 'ACC':
        # Accelerometer data is 3-axis, create three columns
        wrist_dataframes[sensor_name] = pd.DataFrame(data_array, columns=['ACC_x', 'ACC_y', 'ACC_z'])
    else:
        # Other sensors are typically 1D or 2D with one column, flatten and create a DataFrame
        wrist_dataframes[sensor_name] = pd.DataFrame(data_array.flatten(), columns=[sensor_name])

# Load the S2_quest.csv file into a pandas DataFrame
df_s2_quest = pd.read_csv(f'{s2_data_path}/S2_quest.csv')

print("Chest DataFrames (first 5 rows of each):")
for sensor_name, df in chest_dataframes.items():
    print(f"\n{sensor_name}:")
    print(df.head())

print("\nWrist DataFrames (first 5 rows of each):")
for sensor_name, df in wrist_dataframes.items():
    print(f"\n{sensor_name}:")
    print(df.head())

print("\ndf_s2_quest head:")
print(df_s2_quest.head())

Chest DataFrames (first 5 rows of each):

ACC:
    ACC_x   ACC_y   ACC_z
0  0.9554 -0.2220 -0.5580
1  0.9258 -0.2216 -0.5538
2  0.9082 -0.2196 -0.5392
3  0.8974 -0.2102 -0.5122
4  0.8882 -0.2036 -0.4824

ECG:
        ECG
0  0.021423
1  0.020325
2  0.016525
3  0.016708
4  0.011673

EMG:
        EMG
0 -0.004440
1  0.004349
2  0.005173
3  0.007187
4 -0.015152

EDA:
        EDA
0  5.250549
1  5.267334
2  5.243301
3  5.249405
4  5.286407

Temp:
        Temp
0  30.120758
1  30.129517
2  30.138214
3  30.129517
4  30.130951

Resp:
       Resp
0 -1.148987
1 -1.124573
2 -1.152039
3 -1.158142
4 -1.161194

Wrist DataFrames (first 5 rows of each):

ACC:
   ACC_x  ACC_y  ACC_z
0   62.0  -21.0  107.0
1   66.0   13.0   53.0
2   41.0    9.0   15.0
3   52.0   16.0   24.0
4   54.0   15.0   34.0

BVP:
     BVP
0 -59.37
1 -53.42
2 -44.40
3 -33.17
4 -20.79

EDA:
        EDA
0  1.138257
1  1.125444
2  1.011405
3  1.033188
4  0.935807

TEMP:
    TEMP
0  35.41
1  35.41
2  35.41
3  35.41
4  35.41

df_s2_quest h

## Initial Data Inspection

### Subtask:
Perform an initial inspection of the loaded data, including checking the first few rows (.head()), data types (.info()), and descriptive statistics (.describe()) to identify potential issues.


**Reasoning**:
To perform an initial inspection, I will iterate through the `chest_dataframes` and `wrist_dataframes` dictionaries, applying `.head()`, `.info()`, and `.describe()` to each sensor DataFrame. I will also apply these methods to the `df_s2_quest` DataFrame.



In [18]:
print("--- Initial Data Inspection for Chest Sensor DataFrames ---")
for sensor_name, df in chest_dataframes.items():
    print(f"\n----- {sensor_name} Data (Chest) -----")
    print(f"\n{sensor_name}.head():")
    print(df.head())
    print(f"\n{sensor_name}.info():")
    df.info()
    print(f"\n{sensor_name}.describe():")
    print(df.describe())

print("\n--- Initial Data Inspection for Wrist Sensor DataFrames ---")
for sensor_name, df in wrist_dataframes.items():
    print(f"\n----- {sensor_name} Data (Wrist) -----")
    print(f"\n{sensor_name}.head():")
    print(df.head())
    print(f"\n{sensor_name}.info():")
    df.info()
    print(f"\n{sensor_name}.describe():")
    print(df.describe())

print("\n--- Initial Data Inspection for Questionnaire Data (df_s2_quest) ---")
print("\ndf_s2_quest.head():")
print(df_s2_quest.head())
print("\ndf_s2_quest.info():")
df_s2_quest.info()
print("\ndf_s2_quest.describe():")
print(df_s2_quest.describe())

--- Initial Data Inspection for Chest Sensor DataFrames ---

----- ACC Data (Chest) -----

ACC.head():
    ACC_x   ACC_y   ACC_z
0  0.9554 -0.2220 -0.5580
1  0.9258 -0.2216 -0.5538
2  0.9082 -0.2196 -0.5392
3  0.8974 -0.2102 -0.5122
4  0.8882 -0.2036 -0.4824

ACC.info():
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4255300 entries, 0 to 4255299
Data columns (total 3 columns):
 #   Column  Dtype  
---  ------  -----  
 0   ACC_x   float64
 1   ACC_y   float64
 2   ACC_z   float64
dtypes: float64(3)
memory usage: 97.4 MB

ACC.describe():
              ACC_x         ACC_y         ACC_z
count  4.255300e+06  4.255300e+06  4.255300e+06
mean   7.538656e-01 -1.331884e-01 -4.226812e-01
std    1.751162e-01  6.417297e-02  3.002357e-01
min    2.150000e-01 -6.600000e-01 -1.135400e+00
25%    6.210001e-01 -1.498000e-01 -7.172000e-01
50%    8.568000e-01 -1.256000e-01 -3.222000e-01
75%    8.922000e-01 -9.579998e-02 -2.098000e-01
max    2.029800e+00  5.390000e-01  1.246800e+00

----- ECG Data (Ches

## Handle Missing Values

### Subtask:
Identify and address any missing values in the dataset. This may involve imputation, deletion of rows/columns, or other appropriate strategies based on the extent and nature of missing data.


**Reasoning**:
To identify missing values, I will iterate through each DataFrame in the `chest_dataframes` and `wrist_dataframes` dictionaries and print the sum of null values for each column. I will also add a markdown block to acknowledge that `df_s2_quest` needs parsing before checking for missing values.



In [19]:
print("--- Checking for Missing Values in Chest Sensor DataFrames ---")
for sensor_name, df in chest_dataframes.items():
    print(f"\nMissing values in {sensor_name} (Chest):")
    print(df.isnull().sum())

print("\n--- Checking for Missing Values in Wrist Sensor DataFrames ---")
for sensor_name, df in wrist_dataframes.items():
    print(f"\nMissing values in {sensor_name} (Wrist):")
    print(df.isnull().sum())


--- Checking for Missing Values in Chest Sensor DataFrames ---

Missing values in ACC (Chest):
ACC_x    0
ACC_y    0
ACC_z    0
dtype: int64

Missing values in ECG (Chest):
ECG    0
dtype: int64

Missing values in EMG (Chest):
EMG    0
dtype: int64

Missing values in EDA (Chest):
EDA    0
dtype: int64

Missing values in Temp (Chest):
Temp    0
dtype: int64

Missing values in Resp (Chest):
Resp    0
dtype: int64

--- Checking for Missing Values in Wrist Sensor DataFrames ---

Missing values in ACC (Wrist):
ACC_x    0
ACC_y    0
ACC_z    0
dtype: int64

Missing values in BVP (Wrist):
BVP    0
dtype: int64

Missing values in EDA (Wrist):
EDA    0
dtype: int64

Missing values in TEMP (Wrist):
TEMP    0
dtype: int64


```markdown
**Regarding `df_s2_quest`:**

The `df_s2_quest` DataFrame is not yet in a tabular format suitable for a direct `.isnull().sum()` check. It contains metadata and questionnaire responses within a few rows, separated by semicolons. This DataFrame will require further parsing and cleaning before a meaningful assessment of missing values can be performed for individual questions/fields.

Based on the current state of the chest and wrist sensor data, there are no missing values detected.
```

## Handle Duplicates

### Subtask:
Check for and remove any duplicate rows in the dataset to ensure data integrity.


**Reasoning**:
To identify duplicate rows, I will iterate through each DataFrame in the `chest_dataframes` and `wrist_dataframes` dictionaries and use the `.duplicated().sum()` method to count and print the number of duplicate rows for each sensor.



In [20]:
print("--- Checking for Duplicate Rows in Chest Sensor DataFrames ---")
for sensor_name, df in chest_dataframes.items():
    print(f"\nDuplicate rows in {sensor_name} (Chest): {df.duplicated().sum()}")

print("\n--- Checking for Duplicate Rows in Wrist Sensor DataFrames ---")
for sensor_name, df in wrist_dataframes.items():
    print(f"\nDuplicate rows in {sensor_name} (Wrist): {df.duplicated().sum()}")

--- Checking for Duplicate Rows in Chest Sensor DataFrames ---

Duplicate rows in ACC (Chest): 470641

Duplicate rows in ECG (Chest): 4214034

Duplicate rows in EMG (Chest): 4250976

Duplicate rows in EDA (Chest): 4236717

Duplicate rows in Temp (Chest): 4251080

Duplicate rows in Resp (Chest): 4228466

--- Checking for Duplicate Rows in Wrist Sensor DataFrames ---

Duplicate rows in ACC (Wrist): 149038

Duplicate rows in BVP (Wrist): 343616

Duplicate rows in EDA (Wrist): 22615

Duplicate rows in TEMP (Wrist): 24133


```markdown
**Regarding `df_s2_quest`:**

The `df_s2_quest` DataFrame is not yet in a tabular format suitable for a direct `.duplicated().sum()` check. It contains metadata and questionnaire responses within a few rows, separated by semicolons. This DataFrame will require further parsing and cleaning before a meaningful assessment of duplicate rows can be performed for individual questions/fields.

**Regarding Sensor DataFrames:**

Many of the chest and wrist sensor DataFrames contain a significant number of duplicate rows. This will need to be addressed in the next step to ensure data integrity and prevent skewed analysis.
```

**Regarding `df_s2_quest`:**

The `df_s2_quest` DataFrame is not yet in a tabular format suitable for a direct `.duplicated().sum()` check. It contains metadata and questionnaire responses within a few rows, separated by semicolons. This DataFrame will require further parsing and cleaning before a meaningful assessment of duplicate rows can be performed for individual questions/fields.

**Regarding Sensor DataFrames:**

Many of the chest and wrist sensor DataFrames contain a significant number of duplicate rows. This will need to be addressed in the next step to ensure data integrity and prevent skewed analysis.

**Reasoning**:
Many sensor dataframes were found to contain duplicate rows, which can affect data integrity. To address this, I will iterate through each DataFrame in the `chest_dataframes` and `wrist_dataframes` dictionaries and use the `.drop_duplicates()` method to remove all duplicate rows. After removal, I will confirm by re-checking the count of duplicate rows for each DataFrame.



In [21]:
print("--- Removing Duplicate Rows from Chest Sensor DataFrames ---")
for sensor_name, df in chest_dataframes.items():
    initial_duplicates = df.duplicated().sum()
    if initial_duplicates > 0:
        chest_dataframes[sensor_name] = df.drop_duplicates().reset_index(drop=True)
        print(f"Removed {initial_duplicates} duplicate rows from {sensor_name} (Chest). New count: {chest_dataframes[sensor_name].duplicated().sum()}")
    else:
        print(f"No duplicate rows found in {sensor_name} (Chest).")

print("\n--- Removing Duplicate Rows from Wrist Sensor DataFrames ---")
for sensor_name, df in wrist_dataframes.items():
    initial_duplicates = df.duplicated().sum()
    if initial_duplicates > 0:
        wrist_dataframes[sensor_name] = df.drop_duplicates().reset_index(drop=True)
        print(f"Removed {initial_duplicates} duplicate rows from {sensor_name} (Wrist). New count: {wrist_dataframes[sensor_name].duplicated().sum()}")
    else:
        print(f"No duplicate rows found in {sensor_name} (Wrist).")

--- Removing Duplicate Rows from Chest Sensor DataFrames ---
Removed 470641 duplicate rows from ACC (Chest). New count: 0
Removed 4214034 duplicate rows from ECG (Chest). New count: 0
Removed 4250976 duplicate rows from EMG (Chest). New count: 0
Removed 4236717 duplicate rows from EDA (Chest). New count: 0
Removed 4251080 duplicate rows from Temp (Chest). New count: 0
Removed 4228466 duplicate rows from Resp (Chest). New count: 0

--- Removing Duplicate Rows from Wrist Sensor DataFrames ---
Removed 149038 duplicate rows from ACC (Wrist). New count: 0
Removed 343616 duplicate rows from BVP (Wrist). New count: 0
Removed 22615 duplicate rows from EDA (Wrist). New count: 0
Removed 24133 duplicate rows from TEMP (Wrist). New count: 0


## Check Data Types

### Subtask:
Verify that all columns have appropriate data types (e.g., numeric, categorical, datetime) and convert them if necessary to ensure correct processing in subsequent analysis.


**Reasoning**:
To verify data types, I will iterate through each DataFrame in `chest_dataframes` and `wrist_dataframes` and print their data types and information using `.info()`. I will also print the `.info()` for `df_s2_quest` and acknowledge that it requires parsing.



In [22]:
print("--- Checking Data Types for Chest Sensor DataFrames ---")
for sensor_name, df in chest_dataframes.items():
    print(f"\n----- {sensor_name} Data (Chest) -----")
    df.info()

print("\n--- Checking Data Types for Wrist Sensor DataFrames ---")
for sensor_name, df in wrist_dataframes.items():
    print(f"\n----- {sensor_name} Data (Wrist) -----")
    df.info()

print("\n--- Checking Data Types for Questionnaire Data (df_s2_quest) ---")
df_s2_quest.info()

--- Checking Data Types for Chest Sensor DataFrames ---

----- ACC Data (Chest) -----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3784659 entries, 0 to 3784658
Data columns (total 3 columns):
 #   Column  Dtype  
---  ------  -----  
 0   ACC_x   float64
 1   ACC_y   float64
 2   ACC_z   float64
dtypes: float64(3)
memory usage: 86.6 MB

----- ECG Data (Chest) -----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41266 entries, 0 to 41265
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ECG     41266 non-null  float64
dtypes: float64(1)
memory usage: 322.5 KB

----- EMG Data (Chest) -----
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4324 entries, 0 to 4323
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   EMG     4324 non-null   float64
dtypes: float64(1)
memory usage: 33.9 KB

----- EDA Data (Chest) -----
<class 'pandas.core.frame.DataFrame'>
RangeInde

## Final Task

### Subtask:
Summarize the data cleaning steps performed and prepare the cleaned data for further analysis or modeling.


## Summary:

### Data Analysis Key Findings

*   **Data Structure Identification**: The WESAD dataset is organized into subject-specific subdirectories (e.g., `S2/`), each containing a `.pkl` file for processed sensor data, a `_quest.csv` file for questionnaire data, and a `_E4_Data.zip` file for raw data.
*   **Data Loading Strategy**: Due to varying sampling rates and data structures within the `S2.pkl` file, individual pandas DataFrames were created for each chest and wrist sensor (e.g., `ACC`, `ECG`, `BVP`, `EDA`, `Temp`). Accelerometer data was specifically handled to separate its x, y, and z components into distinct columns. The `S2_quest.csv` file was loaded into a separate DataFrame.
*   **Sensor Data Consistency**:
    *   **Chest Sensors**: All chest sensor DataFrames (`ACC`, `ECG`, `EMG`, `EDA`, `Temp`, `Resp`) have a consistent length of 4,255,300 entries, indicating uniform sampling and duration. Their data types are primarily `float64`, with `Temp` being `float32`.
    *   **Wrist Sensors**: Wrist sensor DataFrames show varying lengths, suggesting different sampling rates or recording durations (e.g., `ACC` with 194,528 entries, `BVP` with 389,056 entries, `EDA` and `TEMP` with 24,316 entries). All wrist sensor data are `float64`.
*   **Missing Values**: No missing values were found in any of the chest or wrist sensor DataFrames.
*   **Duplicate Rows**: A significant number of duplicate rows were initially identified across all sensor DataFrames:
    *   Chest sensors: Ranged from 470,641 (ACC) to 4,251,080 (Temp).
    *   Wrist sensors: Ranged from 22,615 (EDA) to 343,616 (BVP).
    *   All identified duplicate rows were successfully removed, resulting in zero duplicates in the cleaned sensor DataFrames.
*   **Questionnaire Data (`df_s2_quest`) State**: The `df_s2_quest` DataFrame was loaded as a single `object` type column. It is not yet in a tabular format, requiring further parsing to extract meaningful questionnaire data and to perform checks for missing values or duplicates within its content.

### Insights or Next Steps

*   **Parse Questionnaire Data**: The `df_s2_quest` DataFrame needs to be parsed and transformed into a structured tabular format to make its contents accessible for analysis and to properly assess its data quality (e.g., missing values, data types).
*   **Synchronize Sensor Data**: Given the different sampling rates and lengths of the wrist sensor DataFrames (and potentially between chest and wrist), future steps should focus on synchronizing these time-series datasets, possibly through resampling or alignment, to enable integrated analysis.


## Save Cleaned Data

### Subtask:
Save the cleaned `df_event_timings`, `df_questionnaire_responses`, and the individual sensor DataFrames (from `chest_dataframes` and `wrist_dataframes`) to CSV files for future use. This will involve creating a new directory for cleaned data and then exporting each DataFrame.

**Reasoning**:
To save the cleaned data, I will first define a path for a new 'cleaned_data' directory within the `S2` subject folder. I will then create this directory if it doesn't already exist. Afterward, I will iterate through the `chest_dataframes` and `wrist_dataframes` dictionaries, saving each sensor DataFrame as a CSV file. Finally, I will save `df_event_timings` and `df_questionnaire_responses` as separate CSV files to the same directory, ensuring all processed data is persistently stored.

In [23]:
import os

# Define the directory to save cleaned data
cleaned_data_dir = f'{s2_data_path}/cleaned_data'

# Create the directory if it does not exist
os.makedirs(cleaned_data_dir, exist_ok=True)

print(f"Saving cleaned data to: {cleaned_data_dir}")

# Save df_event_timings
df_event_timings.to_csv(f'{cleaned_data_dir}/df_event_timings.csv', index=False)
print("Saved df_event_timings.csv")

# Save df_questionnaire_responses
df_questionnaire_responses.to_csv(f'{cleaned_data_dir}/df_questionnaire_responses.csv', index=False)
print("Saved df_questionnaire_responses.csv")

# Save individual chest sensor DataFrames
for sensor_name, df in chest_dataframes.items():
    df.to_csv(f'{cleaned_data_dir}/chest_{sensor_name.lower()}.csv', index=False)
    print(f"Saved chest_{sensor_name.lower()}.csv")

# Save individual wrist sensor DataFrames
for sensor_name, df in wrist_dataframes.items():
    df.to_csv(f'{cleaned_data_dir}/wrist_{sensor_name.lower()}.csv', index=False)
    print(f"Saved wrist_{sensor_name.lower()}.csv")

print("All cleaned DataFrames saved successfully!")

Saving cleaned data to: /content/WESAD_data/WESAD/S2/cleaned_data


NameError: name 'df_event_timings' is not defined

# Task
Inspect the `df_s2_quest` DataFrame to understand its structure for parsing, including printing additional rows and specific slices to identify patterns in the header information and data rows.

## Inspect df_s2_quest structure

### Subtask:
Further inspect the `df_s2_quest` to understand its exact structure and identify patterns for parsing, especially header information and data rows. This may involve printing more rows or specific slices of the DataFrame.


**Reasoning**:
To understand the structure of `df_s2_quest` for parsing, I will display all its rows, then specifically print the first row and a few other distinct rows to identify patterns in how different types of information are presented.



In [24]:
print("--- Displaying all rows of df_s2_quest ---")
print(df_s2_quest.to_string())

print("\n--- First row of df_s2_quest ---")
print(df_s2_quest.iloc[0])

print("\n--- Second row of df_s2_quest ---")
print(df_s2_quest.iloc[1])

print("\n--- Third row of df_s2_quest ---")
print(df_s2_quest.iloc[2])

print("\n--- Fifth row of df_s2_quest (index 4) ---")
print(df_s2_quest.iloc[4])

--- Displaying all rows of df_s2_quest ---
                                     # Subj;S2;;;;;;;;;;;;;;;;;;;;;;;;;
0    # ORDER;Base;TSST;Medi 1;Fun;Medi 2;sRead;fRead;;;;;;;;;;;;;;;;;;;
1   # START;7.08;39.55;70.19;81.25;93.38;54.42;89.51;;;;;;;;;;;;;;;;;;;
2     # END;26.32;50.3;77.1;87.47;100.15;56.07;91.15;;;;;;;;;;;;;;;;;;;
3                                            ;;;;;;;;;;;;;;;;;;;;;;;;;;
4             # PANAS;1;1;3;2;1;3;1;1;1;2;2;2;2;1;4;3;4;4;2;2;2;1;2;1;;
5           # PANAS;3;2;4;1;3;3;1;2;1;4;2;4;3;1;5;4;4;4;2;3;3;3;2;1;3;1
6             # PANAS;1;1;2;3;1;2;1;1;1;1;1;1;3;1;2;1;2;3;1;1;1;1;4;1;;
7             # PANAS;1;1;2;3;1;1;1;1;1;1;1;1;2;1;4;1;1;3;1;1;1;2;3;1;;
8             # PANAS;1;1;1;2;1;1;1;1;1;1;1;1;2;1;2;1;1;2;1;1;1;1;3;1;;
9                                            ;;;;;;;;;;;;;;;;;;;;;;;;;;
10                               # STAI;3;2;1;4;2;3;;;;;;;;;;;;;;;;;;;;
11                               # STAI;1;3;2;1;3;1;;;;;;;;;;;;;;;;;;;;
12                   

The inspection of `df_s2_quest` has provided a clear understanding of its structure:

*   **Single Column Format**: The entire CSV content is loaded into a single column, with each row being a string containing semicolon-separated values.
*   **Header Rows**: The first few rows (`#ORDER`, `#START`, `#END`) provide metadata about the experimental setup (e.g., order of tasks, start and end times). These rows contain both text labels and numeric values.
*   **Questionnaire Data Rows**: Subsequent rows (`#PANAS`, `#STAI`, `#DIM`, `#SSSQ`) contain responses to different questionnaires. Each of these lines starts with a tag (e.g., `#PANAS`) followed by a series of numeric responses, also semicolon-separated.
*   **Empty Rows**: There are rows containing only semicolons (`;;;;...`) which act as separators between different sections of the questionnaire data.
*   **Delimiter**: The primary delimiter for all values appears to be a semicolon (`;`).
*   **Subject Identifier**: The very first line `"# Subj;S2;;;;;;;;;;;;;;;;;;;;;;;;;"` identifies the subject. This information was implicitly captured when loading the `S2_quest.csv` file.

**Next Steps for Parsing `df_s2_quest`:**

1.  **Extract Metadata**: Parse the `#ORDER`, `#START`, and `#END` rows to extract the task sequence and their corresponding start/end times. This data could be stored in a dictionary or a separate DataFrame.
2.  **Extract Questionnaire Data**: Iterate through the remaining rows, identify each questionnaire type (PANAS, STAI, DIM, SSSQ), and extract the numerical responses into separate pandas DataFrames, possibly with appropriate column names derived from the questionnaire structure (if available in documentation or inferable).
3.  **Clean and Convert Data Types**: Ensure extracted numerical data is converted to appropriate numeric types (e.g., `int` or `float`).
4.  **Handle Empty/Irrelevant Values**: Remove or appropriately handle the trailing empty fields (`;;`) and the empty separator rows.

This detailed understanding of the `df_s2_quest` structure will guide the next data cleaning steps to transform this raw questionnaire data into a usable tabular format.

## Extract event timings

### Subtask:
Parse the 'START' and 'END' rows from `df_s2_quest` to extract the event timings for different experimental phases (e.g., Base, TSST, Medi 1, etc.) and store them in a suitable data structure, like a dictionary or a small DataFrame.


**Reasoning**:
To extract event timings, I will locate the '# ORDER', '# START', and '# END' rows, split their string content by semicolons, clean the resulting lists by removing empty strings and the initial '#' character, create event names from the '# ORDER' row, and then construct dictionaries for 'START' and 'END' times. Finally, I will combine these into a new pandas DataFrame as instructed.



In [25]:
import pandas as pd

# 1. Locate and extract the string content from the relevant rows
order_str = df_s2_quest.iloc[0, 0]
start_str = df_s2_quest.iloc[1, 0]
end_str = df_s2_quest.iloc[2, 0]

# 2. Split by semicolon and clean the lists
# Remove empty strings and the initial '#' from the first element

def clean_split_list(s):
    # Remove the first '#' if present and split
    parts = s.replace('#', '').split(';')
    # Filter out empty strings
    return [p.strip() for p in parts if p.strip()]

cleaned_order = clean_split_list(order_str)
cleaned_start = clean_split_list(start_str)
cleaned_end = clean_split_list(end_str)

# 3. Create a list of event names (skipping the 'ORDER' label)
event_names = cleaned_order[1:]

# 4. Create dictionaries for START and END times, converting to float
start_times = {}
for i, event in enumerate(event_names):
    # Ensure index is within bounds for start_str data, skipping the 'START' label
    if (i + 1) < len(cleaned_start):
        try:
            start_times[event] = float(cleaned_start[i + 1])
        except ValueError:
            start_times[event] = None # Handle cases where conversion to float fails

end_times = {}
for i, event in enumerate(event_names):
    # Ensure index is within bounds for end_str data, skipping the 'END' label
    if (i + 1) < len(cleaned_end):
        try:
            end_times[event] = float(cleaned_end[i + 1])
        except ValueError:
            end_times[event] = None # Handle cases where conversion to float fails

# 5. Combine into a new pandas DataFrame
# Create lists for DataFrame construction
events_list = []
start_time_list = []
end_time_list = []

for event in event_names:
    events_list.append(event)
    start_time_list.append(start_times.get(event))
    end_time_list.append(end_times.get(event))

df_event_timings = pd.DataFrame({
    'Event': events_list,
    'Start_Time': start_time_list,
    'End_Time': end_time_list
})

print("Event Timings DataFrame:")
print(df_event_timings)


Event Timings DataFrame:
    Event  Start_Time  End_Time
0    Base        7.08     26.32
1    TSST       39.55     50.30
2  Medi 1       70.19     77.10
3     Fun       81.25     87.47
4  Medi 2       93.38    100.15
5   sRead       54.42     56.07
6   fRead       89.51     91.15


## Extract questionnaire responses

### Subtask:
Parse the 'PANAS', 'STAI', 'DIM', and 'SSSQ' questionnaire response rows from `df_s2_quest`.


**Reasoning**:
To parse the questionnaire responses, I will iterate through the `df_s2_quest` DataFrame, identify rows corresponding to 'PANAS', 'STAI', 'DIM', and 'SSSQ' questionnaires, extract the numeric data by splitting the strings and converting values to float, and then store these responses into separate pandas DataFrames as instructed.



In [26]:
import pandas as pd

# Initialize empty lists to store parsed responses for each questionnaire type
panas_data = []
stai_data = []
dim_data = []
sssq_data = []

# Iterate through df_s2_quest from row index 4 onwards (as metadata is already extracted)
for i in range(4, len(df_s2_quest)):
    row_string = df_s2_quest.iloc[i, 0]

    # Check if the row starts with a questionnaire identifier and process it
    if row_string.startswith('# PANAS'):
        # Remove the tag and split by semicolon, then convert to numeric
        parts = row_string.replace('# PANAS;', '').split(';')
        numeric_parts = [float(p.strip()) for p in parts if p.strip()]
        if numeric_parts:
            panas_data.append(numeric_parts)
    elif row_string.startswith('# STAI'):
        parts = row_string.replace('# STAI;', '').split(';')
        numeric_parts = [float(p.strip()) for p in parts if p.strip()]
        if numeric_parts:
            stai_data.append(numeric_parts)
    elif row_string.startswith('# DIM'):
        parts = row_string.replace('# DIM;', '').split(';')
        numeric_parts = [float(p.strip()) for p in parts if p.strip()]
        if numeric_parts:
            dim_data.append(numeric_parts)
    elif row_string.startswith('# SSSQ'):
        parts = row_string.replace('# SSSQ;', '').split(';')
        numeric_parts = [float(p.strip()) for p in parts if p.strip()]
        if numeric_parts:
            sssq_data.append(numeric_parts)

# Convert lists of lists to pandas DataFrames. pd.DataFrame handles varying row lengths by padding with NaN.
df_panas = pd.DataFrame(panas_data)
df_stai = pd.DataFrame(stai_data)
df_dim = pd.DataFrame(dim_data)
df_sssq = pd.DataFrame(sssq_data)

# Print the head of each created DataFrame to inspect the parsed questionnaire responses
print("\n--- df_panas head ---")
print(df_panas.head())

print("\n--- df_stai head ---")
print(df_stai.head())

print("\n--- df_dim head ---")
print(df_dim.head())

print("\n--- df_sssq head ---")
print(df_sssq.head())


--- df_panas head ---
    0    1    2    3    4    5    6    7    8    9   ...   16   17   18   19  \
0  1.0  1.0  3.0  2.0  1.0  3.0  1.0  1.0  1.0  2.0  ...  4.0  4.0  2.0  2.0   
1  3.0  2.0  4.0  1.0  3.0  3.0  1.0  2.0  1.0  4.0  ...  4.0  4.0  2.0  3.0   
2  1.0  1.0  2.0  3.0  1.0  2.0  1.0  1.0  1.0  1.0  ...  2.0  3.0  1.0  1.0   
3  1.0  1.0  2.0  3.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  3.0  1.0  1.0   
4  1.0  1.0  1.0  2.0  1.0  1.0  1.0  1.0  1.0  1.0  ...  1.0  2.0  1.0  1.0   

    20   21   22   23   24   25  
0  2.0  1.0  2.0  1.0  NaN  NaN  
1  3.0  3.0  2.0  1.0  3.0  1.0  
2  1.0  1.0  4.0  1.0  NaN  NaN  
3  1.0  2.0  3.0  1.0  NaN  NaN  
4  1.0  1.0  3.0  1.0  NaN  NaN  

[5 rows x 26 columns]

--- df_stai head ---
     0    1    2    3    4    5
0  3.0  2.0  1.0  4.0  2.0  3.0
1  1.0  3.0  2.0  1.0  3.0  1.0
2  4.0  1.0  1.0  4.0  2.0  4.0
3  3.0  1.0  1.0  3.0  1.0  4.0
4  4.0  1.0  1.0  4.0  1.0  4.0

--- df_dim head ---
     0    1
0  7.0  2.0
1  5.0  4.

## Create structured DataFrame

### Subtask:
Combine the extracted questionnaire responses into a single, comprehensive pandas DataFrame, ensuring each question's response is in its own column and rows represent individual entries or participants, as appropriate.


**Reasoning**:
To combine the extracted questionnaire responses into a single DataFrame as instructed, I will first rename the columns of each individual questionnaire DataFrame with a descriptive prefix. Then, I will horizontally concatenate 'df_panas', 'df_stai', and 'df_dim', extend 'df_sssq' to match the row count, and finally concatenate it to the combined DataFrame, before printing the head and info of the resulting 'df_questionnaire_responses'.



In [27]:
import numpy as np

# 1. Rename the columns of each DataFrame
df_panas.columns = [f'PANAS_{i}' for i in range(len(df_panas.columns))]
df_stai.columns = [f'STAI_{i}' for i in range(len(df_stai.columns))]
df_dim.columns = [f'DIM_{i}' for i in range(len(df_dim.columns))]
df_sssq.columns = [f'SSSQ_{i}' for i in range(len(df_sssq.columns))]

# 2. Concatenate df_panas, df_stai, and df_dim horizontally
df_combined_responses = pd.concat([df_panas, df_stai, df_dim], axis=1)

# 3. Create a new DataFrame for df_sssq with the same number of rows as df_combined_responses
# Get the number of rows from df_combined_responses
num_rows = len(df_combined_responses)

# Create an empty DataFrame with the correct number of rows and columns from df_sssq
df_sssq_extended = pd.DataFrame(np.nan, index=range(num_rows), columns=df_sssq.columns)

# Place the actual df_sssq data (which is a single row) into the first row of the extended DataFrame
if not df_sssq.empty:
    df_sssq_extended.iloc[0] = df_sssq.iloc[0]

# 4. Horizontally concatenate the extended df_sssq DataFrame with df_combined_responses
df_questionnaire_responses = pd.concat([df_combined_responses, df_sssq_extended], axis=1)

# 5. Print the head and information of the final df_questionnaire_responses DataFrame
print("--- df_questionnaire_responses head ---")
print(df_questionnaire_responses.head())

print("\n--- df_questionnaire_responses info ---")
df_questionnaire_responses.info()

--- df_questionnaire_responses head ---
   PANAS_0  PANAS_1  PANAS_2  PANAS_3  PANAS_4  PANAS_5  PANAS_6  PANAS_7  \
0      1.0      1.0      3.0      2.0      1.0      3.0      1.0      1.0   
1      3.0      2.0      4.0      1.0      3.0      3.0      1.0      2.0   
2      1.0      1.0      2.0      3.0      1.0      2.0      1.0      1.0   
3      1.0      1.0      2.0      3.0      1.0      1.0      1.0      1.0   
4      1.0      1.0      1.0      2.0      1.0      1.0      1.0      1.0   

   PANAS_8  PANAS_9  ...  STAI_4  STAI_5  DIM_0  DIM_1  SSSQ_0  SSSQ_1  \
0      1.0      2.0  ...     2.0     3.0    7.0    2.0     5.0     5.0   
1      1.0      4.0  ...     3.0     1.0    5.0    4.0     NaN     NaN   
2      1.0      1.0  ...     2.0     4.0    7.0    2.0     NaN     NaN   
3      1.0      1.0  ...     1.0     4.0    8.0    1.0     NaN     NaN   
4      1.0      1.0  ...     1.0     4.0    7.0    2.0     NaN     NaN   

   SSSQ_2  SSSQ_3  SSSQ_4  SSSQ_5  
0     4.0     4.

## Final Task

### Subtask:
Summarize the parsing process and the resulting structured DataFrame, including a preview of the cleaned questionnaire data and a brief discussion of its readiness for further analysis.


## Summary:

### Data Analysis Key Findings

*   The initial `df_s2_quest` DataFrame was identified as a single-column structure where each row contained semicolon-separated values.
*   The raw data contained distinct sections: header information (`#ORDER`, `#START`, `#END`), questionnaire responses (`#PANAS`, `#STAI`, `#DIM`, `#SSSQ`), and empty separator rows.
*   Event timings were successfully extracted from the `#ORDER`, `#START`, and `#END` rows into a `df_event_timings` DataFrame. For instance, the 'Base' event has a Start\_Time of 7.08 and an End\_Time of 26.32.
*   Individual questionnaire responses were parsed into separate DataFrames: `df_panas` (26 columns), `df_stai` (6 columns), `df_dim` (2 columns), and `df_sssq` (6 columns), handling varying response lengths with `NaN` values where necessary.
*   The final combined `df_questionnaire_responses` DataFrame consists of 5 entries and 40 columns, with all data converted to `float64`. Columns were systematically renamed (e.g., `PANAS_0`, `STAI_0`), and `df_sssq` (originally a single row) was appropriately extended with `NaN` values to match the row count of other questionnaires during concatenation.

### Insights or Next Steps

*   The questionnaire data is now in a clean, structured format, making it ready for statistical analysis, such as calculating questionnaire scores, conducting correlation analysis with experimental events, or preparing for machine learning models.
*   Further analysis could involve integrating `df_questionnaire_responses` with `df_event_timings` and other subject-specific data to build a holistic dataset for in-depth insights into experimental outcomes and participant responses.


## Verify Saved Files

### Subtask:
List the contents of the `cleaned_data` directory to confirm that all DataFrames were successfully saved as CSV files.

**Reasoning**:
To verify that the files have been saved, I will use the `!ls` command to list the contents of the `cleaned_data_dir`.

In [28]:
print(f"Listing files in {cleaned_data_dir}:")
!ls -F {cleaned_data_dir}

Listing files in /content/WESAD_data/WESAD/S2/cleaned_data:
